lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -1872,6 +1889,131 @@ SILFunction &DifferentiableActivityInfo::getFunction() {
   return parent.function;
 }
 
+void DifferentiableActivityInfo::setVariedAndPropagateToUsers(
+    SILValue value, unsigned independentVariableIndex) {
+  setVaried(value, independentVariableIndex);
+  for (auto *use : value->getUses())
+    propagateVaried(use->getUser(), independentVariableIndex);
+}
+
+void DifferentiableActivityInfo::propagateVaried(
+    SILInstruction *inst, unsigned independentVariableIndex) {
+  // If instruction already propagated variedness, return.
+  if (didPropagateVaried(inst, independentVariableIndex))
+    return;
+  // Propagate variedness for the instruction.
+  //
+  // General rule: if operands are varied, mark results as varied and
+  // recursively propagate variedness to users of results.
+  //
+  // Note: call `setDidPropagateVaried(inst, i)` after propagating variedness if
+  // possible to prevent revisiting instructions that have already propagated
+  // variedness.
+  auto i = independentVariableIndex;
+  // Handle `apply`.
+  if (auto *ai = dyn_cast<ApplyInst>(inst)) {
+    // If callee is non-varying, skip.
+    if (isWithoutDerivative(ai->getCallee()))
+      return;
+    // If any argument is varied, set all direct and indirect results as
+    // varied.
+    for (auto arg : ai->getArgumentsWithoutIndirectResults()) {
+      if (isVaried(arg, i)) {
+        setDidPropagateVaried(ai, i);
+        for (auto indRes : ai->getIndirectSILResults())
+          setVariedAndPropagateToUsers(indRes, i);
```

Comment:
Great catch! I'll preemptively change indirect result propagation to use `propagateVariedInwardsThroughProjections` instead of `setVariedAndPropagateToUsers`, for good measure.

---

In practice, I couldn't think of any Swift code where `apply` has an indirect result that's not a local `alloc_stack`. Here's one I tried:

```swift
struct Tuple: Differentiable {
  var x, y: Float
}

@differentiable
func id<T: Differentiable>(_ x: T) -> T { x }

func foo(_ x: Float) -> Float {
  var tuple = Tuple(x: x, y: x)
  return id(tuple.x) // apply with indirect results
}
print(gradient(at: 3, in: foo))
```

```
[AD] Activity info for $foo at (source=0 parameters=(0))
bb0:
[ACTIVE] %0 = argument of bb0 : $Float                     // users: %5, %5, %1
[ACTIVE]   %2 = alloc_stack $Tuple, var, name "tuple"      // users: %19, %6, %8
[USEFUL]   %3 = metatype $@thin Tuple.Type                 // user: %5
[NONE]   // function_ref Tuple.init(x:y:)
  %4 = function_ref @$s6apply25TupleV1x1yACSf_SftcfC : $@convention(method) (Float, Float, @thin Tuple.Type) -> Tuple // user: %5
[ACTIVE]   %5 = apply %4(%0, %0, %3) : $@convention(method) (Float, Float, @thin Tuple.Type) -> Tuple // user: %6
[ACTIVE]   %7 = alloc_stack $Float                         // users: %18, %17, %15
[ACTIVE]   %8 = begin_access [read] [static] %2 : $*Tuple  // users: %11, %9
[ACTIVE]   %9 = struct_element_addr %8 : $*Tuple, #Tuple.x // user: %10
[ACTIVE]   %10 = load [trivial] %9 : $*Float               // user: %13
[ACTIVE]   %12 = alloc_stack $Float                        // users: %16, %15, %13
[NONE]   // function_ref id<A>(_:)
  %14 = function_ref @$s6apply22idyxxs14DifferentiableRzlF : $@convention(thin) <τ_0_0 where τ_0_0 : Differentiable> (@in_guaranteed τ_0_0) -> @out τ_0_0 // user: %15
[NONE]   %15 = apply %14<Float>(%7, %12) : $@convention(thin) <τ_0_0 where τ_0_0 : Differentiable> (@in_guaranteed τ_0_0) -> @out τ_0_0
[ACTIVE]   %17 = load [trivial] %7 : $*Float               // user: %20
```

`%12 = alloc_stack` is the indirect result of the `apply` instruction.

If anyone has ideas for Swift code where SILGen generates `apply` with an indirect result that’s an address projection, please let me know!

Reply:
TF-947 tracks checking whether `apply` instruction can have an indirect result that's an address projection.